### PR TITLE
refactor: unite supervisor in one subgraph

### PR DESCRIPTION
**Pull request by grischperl**
## Description

Unite the Supervisor together with Finalizer and Planner under one subgraph.
The subgraph is now called the Supervisor, with three nodes: Finalizer, Planner and Router.
The newly introduced Router simply routes the task to the assigned agent.

### Changes proposed in this pull request:

- create Supervisor subgraph
- move Finalizer and Planner into that subgraph
- Improve naming (e.g. "KmyaGraph -> CompanionGraph)
- Fix unit and integration tests
- Add integration tests for Finalizer and Router
- Improve Finalizer prompt to make the integration tests green

supervisor

Also take a look at the following comment for a comprehensive To-Do list implemented in this PR: [Comment #170](https://github.com/kyma-project/kyma-companion/issues/170#issuecomment-2429266175)

### Related issue(s)

#170

### CODE REVIEW

File: src/agents/prompts.py [hunk 1]
```diff
@@ -111,37 +111,23 @@
...
-If you reject it, politely apologize to the user for not being able to answer their query without providing an answer 
-to the user query yourself. Avoid answering the user query yourself. Avoid giving any impression regarding the
-correctness or incorrectness of the rejected responses.
-Remove any information regarding the agents and your decision-making process from your final response.
+You are an expert in Kubernetes, Kyma and common knowledge topics.
+Your task is to review responses from other agents: "{members}" to a specific user query: "{query}".
+Do not answer the user query yourself.
+
+First, decide whether to accept or reject the response.
+Accept the response if it is answers the user query correctly and remove the agent names and your decision from your response.
```
Comment (mfaizanse) [inline src/agents/prompts.py:119]:
```suggestion
Accept the response if it answers the user query correctly and remove the agent names and your decision from your response.
```

File: src/agents/supervisor/agent.py [hunk 2]
```diff
@@ -1,91 +1,298 @@
...
+        gpt_4o = models[LLM.GPT4O]
 
-    def __init__(self, model: IModel, members: list[str]):
-        self.model = model
-        self.options = [FINALIZER] + members
-        self.parser = self._create_parser()
-        self.supervisor_chain = self._create_supervisor_chain()
+        self.model = gpt_4o
+        self.members = members
+        self.options = members
```
Comment (mfaizanse) [inline src/agents/supervisor/agent.py:78]: Whats the difference between `self.members` and `self.options`?

File: src/agents/supervisor/agent.py [hunk 3]
```diff
@@ -1,91 +1,298 @@
...
         """Agent name."""
         return self._name
 
-    def agent_node(self):  # noqa ANN
+    def agent_node(self) -> CompiledGraph:
         """Get Supervisor agent node function."""
+        return self._graph
+
+    def _route(self, state: AgentState) -> dict[str, Any]:
+        """Supervisor node."""
```
Comment (mfaizanse) [inline src/agents/supervisor/agent.py:122]: Can you update this description?

File: src/agents/supervisor/agent.py [hunk 4]
```diff
@@ -1,91 +1,298 @@
...
+                        [subtask.__dict__ for subtask in state.subtasks]
+                    ),
+                },
+            )
+            route_result = self.parser.parse(result.content)
+            return {
+                "next": route_result.next,
+                "subtasks": state.subtasks,
+                "messages": [AIMessage(content=result.content, name=self.name)],
+            }
```
Comment (mfaizanse) [inline src/agents/supervisor/agent.py:137]: Can this same thing be done without using LLM? Cann't we loop over the subtasks and check for next pending subtask? If all subtasks are marked as completed, then route to finalizer?

File: src/agents/supervisor/agent.py [hunk 5]
```diff
@@ -1,91 +1,298 @@
...
+from agents.common.constants import (
+    FINALIZER,
+    MESSAGES,
+    NEXT,
+    PLANNER,
+)
+from agents.common.state import AgentState, Plan
+from agents.common.utils import create_node_output, filter_messages
+from agents.prompts import FINALIZER_PROMPT, PLANNER_PROMPT
+from agents.supervisor.prompts import ROUTER_ROLE_PROMPT, ROUTER_TASK_PROMPT
```
Comment (mfaizanse) [inline src/agents/supervisor/agent.py:23]: Should we move the `FINALIZER_PROMPT`, `PLANNER_PROMPT` to `agents.supervisor.prompts`?

File: src/agents/supervisor/agent.py [hunk 6]
```diff
@@ -1,91 +1,298 @@
...
             try:
-                result = self.supervisor_chain.invoke(
-                    input={
-                        "messages": filter_messages(state.messages),
-                        "subtasks": json.dumps(
-                            [subtask.__dict__ for subtask in state.subtasks]
-                        ),
-                    },
+                plan = self.plan_parser.parse(response_content)
+                if plan.response:
```
Comment (mfaizanse) [inline src/agents/supervisor/agent.py:185]: Can you add a comment here to explain the if condition?

File: src/agents/supervisor/agent.py (hunk 2 above)
Comment (grischperl) [reply src/agents/supervisor/agent.py:78]: Good point, there is none.

File: src/agents/supervisor/agent.py (hunk 3 above)
Comment (grischperl) [reply src/agents/supervisor/agent.py:122]: done

File: src/agents/prompts.py (hunk 1 above)
Comment (grischperl) [reply src/agents/prompts.py:119]: done

File: src/agents/supervisor/agent.py (hunk 5 above)
Comment (grischperl) [reply src/agents/supervisor/agent.py:23]: yes!

File: src/agents/supervisor/agent.py (hunk 4 above)
Comment (grischperl) [reply src/agents/supervisor/agent.py]: you are right, done

**Comment by grischperl** [comment]
##  Evaluation Test results:
### 30.10.24 07:55 PM 
```
************************************************************************
*** Test Results per Category:
************************************************************************

+------------------------+-----------+
|     Category Name      | Score (%) |
+------------------------+-----------+
|       kubernetes       |   16.33   |
|    problem_finding     |   52.63   |
|    solution_finding    |    4.1    |
| definition/explanation |    0.0    |
| step-by-step_guidance  |   15.15   |
|          yaml          |    0.0    |
|      completeness      |    0.0    |
|          kyma          |   13.04   |
+------------------------+-----------+

************************************************************************
*** Overall success score across all expectations: 14.07%
************************************************************************
```

**Review by mfaizanse — APPROVED** [review]

**Pull request merged**